### PR TITLE
Autolink headers

### PR DESCRIPTION
**Pull request by wackerow**
## Description
- Adds anchor links back to markdown headers on hover
- Uses `Icon` component with an icon from `react-icons` that matches our existing design
- Created as a single component within `MdComponents` called `IdAnchor`, along with a function that produces the needed props for the parent heading given an `id`.
- Implemented these within the four shared markdown heading components in `MdComponents`

Included `scrollMarginTop` with these props which removed the need for previous false top margins to allow a link not be hidden by the nav bar. As a result, was able to also remove some of that existing logic on the `Staking` and `Static` layouts, and replace with the shared components (while overriding the custom props that were different). This allowed reusability of the IdAnchor logic for all layouts. 

## Related Issue
https://www.notion.so/efdn/Add-MD-heading-links-9a8610011f814a45b16a4fb7a398cfbb?pvs=4

### CODE REVIEW

**Review by corwintines — APPROVED** [review]

File: src/components/MdComponents.tsx [hunk 1]
```diff
@@ -32,44 +34,73 @@ import InfoBanner from "./InfoBanner"
...
 // import MeetupList from "@/components/MeetupList"
 
 /**
  * Base HTML elements
  */
-export const commonHeadingProps: HeadingProps = {
+const headingPropsForAnchor = (id?: string): HeadingProps =>
+  id
+    ? ({
+        scrollMarginTop: 28,
```
Comment (wackerow) [inline src/components/MdComponents.tsx:48]: Got rid of false top margins and replaced with `scrollMarginTop`... fixes things like this:

https://github.com/ethereum/ethereum-org-fork/assets/54227730/b128f1d6-33aa-4f05-af72-03a4486e1bdc

File: src/components/MdComponents.tsx [hunk 2]
```diff
@@ -32,44 +34,73 @@ import InfoBanner from "./InfoBanner"
...
+  id ? (
+    <Link href={`#${id}`}
+      position="absolute"
+      insetInlineEnd="100%"
+    >
+      <Icon
+        as={CiLink}
+        opacity={0}
+        _groupHover={{ opacity: 1 }}
+        transition="opacity 0.1s ease-in-out"
```
Comment (wackerow) [inline src/components/MdComponents.tsx:72]: Used `opacity` to enable a quick transition effect

File: src/components/MdComponents.tsx [hunk 3]
```diff
@@ -32,44 +34,73 @@ import InfoBanner from "./InfoBanner"
...
+        me={1}
+      />
+    </Link>
+  ) : null
+
+export const Heading1 = ({ children, ...rest }: HeadingProps) => (
+  <OldHeading as="h1" {...commonHeadingProps()} fontSize="2.5rem" {...rest}>
+    {children}
+  </OldHeading>
 )
```
Comment (wackerow) [inline src/components/MdComponents.tsx:83]: Left autolinking off for H1s as these are the top of the page (parity with Gatsby site)

File: src/components/MdComponents.tsx [hunk 4]
```diff
@@ -32,44 +34,73 @@ import InfoBanner from "./InfoBanner"
...
+  ...headingPropsForAnchor(id),
+})
 
-export const Heading1 = (props: HeadingProps) => (
-  <OldHeading as="h1" {...commonHeadingProps} fontSize="2.5rem" {...props} />
+const IdAnchor: React.FC<{ id?: string }> = ({ id }) =>
+  id ? (
+    <Link href={`#${id}`}
+      position="absolute"
+      insetInlineEnd="100%"
```
Comment (wackerow) [inline src/components/MdComponents.tsx:65]: ("right", but direction responsive)

File: src/components/MdComponents.tsx [hunk 5]
```diff
@@ -32,44 +34,73 @@ import InfoBanner from "./InfoBanner"
 import Card from "./Card"
 import QuizWidget from "./Quiz/QuizWidget"
 import DocLink from "./DocLink"
+import Link from "@/components/Link"
+import { motion } from "framer-motion"
```
Comment (corwintines) [inline src/components/MdComponents.tsx]:
```suggestion
```

I dont see this used anywhere

File: src/components/MdComponents.tsx [hunk 6]
```diff
@@ -32,44 +34,73 @@ import InfoBanner from "./InfoBanner"
...
-export const Heading4 = (props: HeadingProps) => (
-  <OldHeading
-    as="h4"
-    {...commonHeadingProps}
-    fontSize="xl"
-    fontWeight={600}
-    {...props}
-  />
+export const Heading4 = ({ id, children, ...rest }: HeadingProps) => (
+  <OldHeading as="h4" {...commonHeadingProps} fontSize="xl" fontWeight={600} {...rest}>
```
Comment (pettinarip) [inline src/components/MdComponents.tsx]:
```suggestion
  <OldHeading as="h4" {...commonHeadingProps(id)} fontSize="xl" fontWeight={600} {...rest}>
```

**Review by pettinarip — COMMENTED** [review]
Great job @wackerow 💪🏼 

Comparing it with prod, I can see that we had an `aria-label` before composed of the `id` + the word "permalink". Not sure how much that is helping the user but do you think it is something we should bring back?
image

File: src/components/MdComponents.tsx (hunk 6 above)
Comment (wackerow) [reply src/components/MdComponents.tsx]: oh good catch, thanks

**Comment by wackerow** [comment]
@pettinarip Aria-labels usually don't hurt, so added this to link props:

```ts
aria-label={id.replaceAll("-", " ") + " permalink"}
```

...which should produce the same as prod:

image

---

Committed the other suggestions from the code reviews... Should be good to merge

**Review by corwintines — APPROVED** [review]

**Review by pettinarip — APPROVED** [review]

**Pull request merged**